src/upf_network.py

```diff
@@ -0,0 +1,210 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Abstract the network configuration for the UPF service."""
+
+import logging
+from typing import List, Optional
+
+import iptc
+from pyroute2 import NDB, IPRoute, NetlinkError
+
+logger = logging.getLogger(__name__)
+
+
+class UPFNetworkError(Exception):
+    """Custom exception for UPFNetwork."""
+
+
+class NetworkInterface:
+    """A class to interact with a network interface."""
+
+    def __init__(self, name: str):
+        self.network_db = NDB()
+        self.ip_route = IPRoute()
+        self.name = name
+
+    def exists(self) -> bool:
+        """Return whether the network interface exists."""
+        return self.name in self.network_db.interfaces  # type: ignore[reportAttributeAccessIssue]
+
+    def is_valid(self) -> bool:
+        """Return whether the network interface is valid."""
+        if not self.exists():
+            logger.warning("Interface %s does not exist", self.name)
+            return False
+        ip_address = self.get_ip_address()
+        if not ip_address:
+            logger.warning("IP address for interface %s is empty", self.name)
+            return False
+        return True
+
+    def get_ip_address(self) -> str:
+        """Get the IP address of the given network interface."""
+        interfaces = self.network_db.interfaces  # type: ignore[reportAttributeAccessIssue]
+        try:
+            iface_record = interfaces[self.name]
+            ip_addresses = iface_record.ipaddr
+            for ip in ip_addresses:
+                if ip.family == 2:
```

Comment:
We can import the constant from the socket module to clarify this magic number:

```
from socket import AF_INET
```

```suggestion
                if ip.family == AF_INET:
```

Reply:
Done